NumLib/ODESolver/NonlinearSolver.cpp

```diff
@@ -215,63 +215,65 @@ bool NonlinearSolver<NonlinearSolverTag::Newton>::solve(
         sys.applyKnownSolutionsNewton(J, res, minus_delta_x);
         INFO("[time] Applying Dirichlet BCs took %g s.", time_dirichlet.elapsed());
 
-        if (!sys.isLinear() && _convergence_criterion->hasResidualCheck())
```

Comment:
without checking the residual, how can the iteration stop? besides, if the residual check is satisfied, there is no need to solve a linear system even if it's a linear problem. checking residual is cheaper than linear solve.